src/infra/core/src/auth/authentication_service_impl.rs

```diff
@@ -0,0 +1,214 @@
+use jsonwebtoken::errors::ErrorKind;
+use jsonwebtoken::{decode, encode, Algorithm, DecodingKey, EncodingKey, Header, Validation};
+use kamu_core::auth::*;
+use rand::distributions::Alphanumeric;
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+
+///////////////////////////////////////////////////////////////////////////////
+
+const KAMU_JWT_ISSUER: &str = "dev.kamu";
+const KAMU_JWT_ALGORITHM: Algorithm = Algorithm::HS384;
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct AuthenticationServiceImpl {
+    kamu_jwt_secret: String,
+    authentication_providers_by_method: HashMap<&'static str, Arc<dyn AuthenticationProvider>>,
+}
+
+#[component(pub)]
+impl AuthenticationServiceImpl {
+    pub fn new(authentication_providers: Vec<Arc<dyn AuthenticationProvider>>) -> Self {
+        let kamu_jwt_secret = match std::env::var("KAMU_JWT_SECRET") {
+            Ok(jwt_secret) => jwt_secret,
+            Err(_) => Self::random_jwt_secret(),
+        };
+
+        let mut service = Self {
+            authentication_providers_by_method: HashMap::new(),
+            kamu_jwt_secret,
+        };
+
+        for authentication_provider in authentication_providers {
+            let login_method = authentication_provider.login_method();
+
+            let insert_result = service
+                .authentication_providers_by_method
+                .insert(login_method, authentication_provider);
+
+            if let Some(_) = insert_result {
+                panic!(
+                    "Duplicate authentication provider for method {}",
+                    login_method
+                );
+            }
+        }
+
+        service
+    }
+
+    fn random_jwt_secret() -> String {
+        thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(64)
+            .map(char::from)
+            .collect()
+    }
+
+    fn resolve_authentication_provider(
+        &self,
+        login_method: &str,
+    ) -> Result<Arc<dyn AuthenticationProvider>, UnsupportedLoginMethodError> {
+        match self.authentication_providers_by_method.get(login_method) {
+            Some(provider) => Ok(provider.clone()),
+            None => Err(UnsupportedLoginMethodError {
+                method: login_method.into(),
+            }),
+        }
+    }
+
+    fn make_access_token(
+        &self,
+        subject: String,
+        login_method: &str,
+        provider_credentials_json: String,
+    ) -> Result<String, InternalError> {
+        let claims = KamuAccessTokenClaims {
+            iat: Utc::now().timestamp() as usize,
+            exp: (Utc::now() + Duration::days(1)).timestamp() as usize,
+            iss: String::from(KAMU_JWT_ISSUER),
+            sub: subject,
+            access_credentials: KamuAccessCredentials {
+                login_method: String::from(login_method),
+                provider_credentials_json,
+            },
+        };
+
+        encode(
```

Comment:
Does this create an encrypted + signed token or only signed?

Reply:
`encode` here sends for Base64 conversion. Then it's signed. But it's not encrypted - there is no sensitive information there anyway.